derivation/derivation.go

```diff
@@ -0,0 +1,63 @@
+// nolint: govet
+package derivation
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+)
+
+type Derivation struct {
+	_       string  `@DerivationPrefix`
+	Content []Value `'('((@@ ','?)* @@?)? ')'`
+}
+
+type Value struct {
+	Outputs          []Output          `'[' ((@@ ','?)* @@? )']'','`
+	InputDerivations []InputDerivation `'[' ((@@ ','?)* @@? )?']'','`
+	InputSources     []string          `'[' ((@NixPath ','?)* @NixPath? )?']'','`
+	Platform         string            `@String ','`
+	Builder          string            `@String ','`
+	Arguments        []string          `'[' ((@String ','?)* (@NixPath|@String)? )?']'','`
+	EnvVars          []Env             `'[' ((@@ ','?)* (@@)? )']'`
+}
+
+type Output struct {
+	Content       string `'(' @String ','`
+	Path          string `@NixPath ','`
+	HashAlgorithm string `@String ','`
+	Hash          string `@String ')'`
+}
+
+type InputDerivation struct {
+	Path string   `'(' @NixPath ','`
+	Name []string `'[' ((@String ','?)* @String? )']' ')'`
+}
+
+type Env struct {
+	Key   string `'(' @String ','`
+	Value string `(@String|@NixPath)? ')'`
+}
+
+func ReadDerivation(filename string) (*Derivation, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err = file.Close(); err != nil {
+			log.Fatal(err)
+		}
+	}()
```

Comment:
Right now, we mostly don't do much error handling when `defer`ing things that can return an error. I'd probably just `defer file.Close()`. However, if we change the whole function to accept a io.Reader (not io.ReadCloser) (and leave closing up to the caller), we don't need any `defer` at all.